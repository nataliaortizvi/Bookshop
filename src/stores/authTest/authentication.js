import { defineStore } from "pinia";
import { signInWithEmailAndPassword } from "firebase/auth";
import { createUserWithEmailAndPassword } from "firebase/auth";
import { signOut } from "firebase/auth";
import { db, auth } from "../../firebase/config";
import { doc, setDoc, updateDoc, onSnapshot, deleteDoc, deleteField } from "firebase/firestore";

///// OPTIONS STORE
export const useAuthenticationStore = defineStore("authentication", {
    state: () => ({
        user: null,
        currentUser: null,
    }),
    actions: {
        //old user
        async signIn(email, password) {
            await signInWithEmailAndPassword(auth, email, password);
        },

        //new user
        async signUp(email, password, name) {

            var ref;
            try {
                const userCredential = await createUserWithEmailAndPassword(auth, email, password);

                ref = userCredential.user;

                this.user = {
                    name: name,
                    email: email,
                    favorites: [],
                    id: ref.uid,
                }
                setDoc(doc(db, "users", ref.uid), this.user);

            } catch (error) {
                //alert(error.message);  
            }
        },

        //log out 
        logOut() {
            signOut(auth);
        },

        //----------------------------USERS DATABASE--------------------------
        getCurrentUser(currentUid) {
            if (currentUid != null) {
                onSnapshot(doc(db, "users", currentUid), docSnapshot => {
                    this.currentUser = docSnapshot.data();
                    //console.log('DOC:', this.currentUser)
                });
            } else {
                return null
            }
        },

        loadCurrentUser() {
            if (this.currentUser != null) {
                return this.currentUser;
            } else {
                return null
            }
        },

        addRating(rate, book) {
            //console.log("rate: ", rate, "book: ", book.id);
            let by;
            let votesArray = book.votes;

            by = this.currentUser.id;

            let vote = {
                by,
                rate
            }

            console.log("CUANTOS VOTOS: ", votesArray);

            if (votesArray.length != 0) {
                console.log("SIIII", votesArray.length);

                let votePosition = votesArray.findIndex(v => v.by == by);
                //console.log("POSICION", votePosition);

                if (votePosition != -1) {
                    console.log("YA VOTO: ", votesArray[votePosition]);

                    //console.log("ARRAY EN EL VOTO", votesArray[votePosition].rate);
                    votesArray[votePosition].rate = rate;

                    updateDoc(doc(db, "products", book.id), {
                        "votes": votesArray
                    });

                } else {
                    console.log("NO HAY VOTO: ", votesArray[votePosition]);

                    votesArray.push(vote)

                    updateDoc(doc(db, "products", book.id), {
                        "votes": votesArray
                    });
                }

            } else {
                console.log("NOOO", votesArray.length);

                votesArray.push(vote)

                updateDoc(doc(db, "products", book.id), {
                    "votes": votesArray
                });
            }
        },

        addFavorites(book) {
            let favorite = this.currentUser.favorites;

            let favBookPosition = favorite.findIndex(v => v.id == book.id);

            if (favBookPosition != -1) {
                //console.log("YA ESTA DE FAVORITO", favBookPosition);

                favorite.splice(favBookPosition, 1)

                updateDoc(doc(db, "users", this.currentUser.id), {
                    "favorites": favorite
                });

            } else {
                favorite.push(book);
                updateDoc(doc(db, "users", this.currentUser.id), {
                    "favorites": favorite
                });
                //console.log("NO ESTA DE FAVORITO", favBookPosition);
            }
        },
    },
});