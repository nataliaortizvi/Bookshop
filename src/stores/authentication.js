import { defineStore } from "pinia";
import { signInWithEmailAndPassword } from "firebase/auth";
import { createUserWithEmailAndPassword } from "firebase/auth";
import { signOut } from "firebase/auth";
import { db, auth } from "../firebase/config";
import { doc, setDoc, updateDoc, onSnapshot, addDoc } from "firebase/firestore";

///// OPTIONS STORE
export const useAuthenticationStore = defineStore("authentication", {
    state: () => ({
        isUser: null,
        user: {},
        currentUser: null,
    }),
    actions: {
        signIn(email, password) {
            signInWithEmailAndPassword(auth, email, password)
                .then((userCredential) => {
                    const user = userCredential.user;
                })
                .catch((error) => {
                    const errorCode = error.code;
                    const errorMessage = error.message;
                    alert('User not found');
                });
        },

        signUp(email, password, name) {
            var ref;
            createUserWithEmailAndPassword(auth, email, password)
                .then((userCredential) => {
                    ref = userCredential.user;
                })
                .then(() => {
                    this.user = {
                        name: name,
                        email: email,
                        favorites: [],
                        id: ref.uid,
                    }
                    setDoc(doc(db, "users", ref.uid), this.user);
                })
                .catch((error) => {
                    const errorCode = error.code;
                    const errorMessage = error.message;
                    alert(errorMessage);
                });
        },

        logOut() {
            signOut(auth).then(() => {
                // Sign-out successful.
            }).catch((error) => {
                // An error happened.
            });
        },

        //----------------------------USERS DATABASE--------------------------
        getCurrentUser(currentUid) {
            if (currentUid != null) {
                onSnapshot(doc(db, "users", currentUid), docSnapshot => {
                    this.currentUser = docSnapshot.data();
                    //console.log('DOC:', this.currentUser)
                });
            } else {
                return null
            }

        },

        loadCurrentUser() {
            if (this.currentUser != null) {
                return this.currentUser;
            } else {
                return null
            }
        },

        addRating(rate, book) {
            //console.log("rate: ", rate, "book: ", book.id);
            let by;
            let votesArray = book.votes;
            let stars;

            by = this.currentUser.id;

            let vote = {
                by,
                rate
            }

            //console.log("CUANTOS VOTOS: ", votesArray);

            if (votesArray.length != 0) {
                console.log("SIIII", votesArray.length);

                let votePosition = votesArray.findIndex(v => v.by == by);
                console.log("POSICION", votePosition);
                
                if (votePosition != -1){
                    console.log("YA VOTO: ", votesArray[votePosition]);

                    console.log("ARRAY EN EL VOTO", votesArray[votePosition].rate);
                    votesArray[votePosition].rate = rate;

                    updateDoc(doc(db, "products", book.id), {
                        "votes": votesArray
                    });
                    
                }else{
                    console.log("NO HAY VOTO: ", votesArray[votePosition]);

                    votesArray.push(vote)

                    updateDoc(doc(db, "products", book.id), {
                        "votes": votesArray
                    });
                }

            } else {
                console.log("NOOO", votesArray.length);

                votesArray.push(vote)

                updateDoc(doc(db, "products", book.id), {
                    "votes": votesArray
                });
            }
        },
    },
});