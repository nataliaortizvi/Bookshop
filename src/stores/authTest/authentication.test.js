import { createPinia, setActivePinia } from "pinia";
import { describe, vi, vitest, expect, beforeEach, it } from "vitest";
import { useAuthenticationStore } from "./authentication";


describe("Auth store", () => {

    vi.mock("firebase/auth", () => {
        const signInWithEmailAndPassword = vitest.fn(() => Promise.resolve({ user: "1234" }));
        const createUserWithEmailAndPassword = vitest.fn(() => Promise.resolve({ user: "1234" }));
        const signOut = vitest.fn(() => Promise.resolve(undefined));

        return {
            signInWithEmailAndPassword,
            createUserWithEmailAndPassword,
            signOut,
        };
    });

    vi.mock("../../firebase/config", () => ({ auth: "mock" }))

    beforeEach(() => {
        setActivePinia(createPinia());
    });


    it("sign Up fn", async () => {
        const authentication = useAuthenticationStore();
        //console.log = vitest.fn();

        await authentication.signUp('email', 'pass', 'name');
       
        expect(authentication.user.name).toEqual('name');
        expect(authentication.user.email).toEqual('email');
    });

    it("sign In fn", async () => {
        const authentication = useAuthenticationStore();

            await authentication.signIn('email', 'pass');

            //await expect(authentication.signIn()).rejects.toThrow(Error);
    });

    it("log out fn", () => {
        const authentication = useAuthenticationStore();

        authentication.logOut();
    });


    it("load current user fn", async () => {
        const authentication = useAuthenticationStore();

        await authentication.signIn('email', 'pass');
        await authentication.loadCurrentUser();

        expect(authentication.currentUse).toBeFalsy();
    });

    it("user is null", () => {
        const authentication = useAuthenticationStore();

        expect(authentication.user).toBeNull();
    });
})