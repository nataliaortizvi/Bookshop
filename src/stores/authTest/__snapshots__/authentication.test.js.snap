// Vitest Snapshot v1

exports[`Auth store > load current user fn 1`] = `null`;
